plugins/node/opentelemetry-koa-instrumentation/src/koa.ts

```diff
@@ -0,0 +1,139 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { BasePlugin } from '@opentelemetry/core';
+import type * as koa from 'koa';
+import * as shimmer from 'shimmer';
+import {
+  KoaMiddleware,
+  KoaContext,
+  KoaComponentName,
+  kLayerPatched,
+} from './types';
+import { VERSION } from './version';
+import { getMiddlewareMetadata } from './utils';
+
+/** Koa instrumentation for OpenTelemetry */
+export class KoaInstrumentation extends BasePlugin<typeof koa> {
+  static readonly component = KoaComponentName;
+  readonly supportedVersions = ['^2.0.0'];
+
+  constructor(readonly moduleName: string) {
+    super('@opentelemetry/koa-instrumentation', VERSION);
+  }
+
+  /**
+   * Patches Koa operations by wrapping the Koa.use function
+   */
+  protected patch(): typeof koa {
+    this._logger.debug('Patching Koa');
+    if (this._moduleExports == null) {
+      return this._moduleExports;
+    }
+    this._logger.debug('Patching Koa.use');
+    shimmer.wrap(this._moduleExports.prototype, 'use', this._getKoaUsePatch);
+
+    return this._moduleExports;
+  }
+
+  /**
+   * Unpatches all Koa operations
+   */
+  protected unpatch(): void {
+    this._logger.debug('Unpatching Koa');
+    shimmer.unwrap(this._moduleExports.prototype, 'use');
+  }
+
+  /**
+   * Patches the Koa.use function in order to instrument each original
+   * middleware layer which is introduced
+   * @param {KoaMiddleware} middleware - the original middleware function
+   */
+  private _getKoaUsePatch(original: (middleware: KoaMiddleware) => koa) {
+    return function use(this: koa, middlewareFunction: KoaMiddleware) {
+      let patchedFunction: KoaMiddleware;
+      if (middlewareFunction.router) {
+        patchedFunction = plugin._patchRouterDispatch(middlewareFunction);
+      } else {
+        patchedFunction = plugin._patchLayer(middlewareFunction, false);
+      }
+      return original.apply(this, [patchedFunction]);
+    };
+  }
+
+  /**
+   * Patches the dispatch function used by @koa/router. This function
+   * goes through each routed middleware and adds instrumentation via a call
+   * to the @function _patchLayer function.
+   * @param {KoaMiddleware} dispatchLayer - the original dispatch function which dispatches
+   * routed middleware
+   */
+  private _patchRouterDispatch(dispatchLayer: KoaMiddleware): KoaMiddleware {
+    this._logger.debug('Patching @koa/router dispatch');
+
+    const router = dispatchLayer.router;
+
+    const routesStack = router?.stack ?? [];
+    for (const pathLayer of routesStack) {
+      const path = pathLayer.path;
+      const pathStack = pathLayer.stack;
+      for (let j = 0; j < pathStack.length; j++) {
+        const routedMiddleware: KoaMiddleware = pathStack[j];
+        pathStack[j] = this._patchLayer(routedMiddleware, true, path);
```

Comment:
Thank you for bringing this up! Both nested and prefixed routers are handled currently and I believe the span names and attributes should function as expected. These routes are instrumented at the time that the routes are actually added to the app itself (in `app.use`).

For example, for the following nested router:
```javascript
const nestedRouter = new Router();
nestedRouter.get('/post/:id', showPost);
router.use('/:first', nestedRouter.routes());
app.use(router.routes());
```
The span recorded for a request to `/testparam/post/2` looks as follows: 
![nestedRouter](https://user-images.githubusercontent.com/60523681/88860172-32416680-d1c9-11ea-875c-2150741b52e4.jpg)

Prefixed routes such as the following are also instrumented the same way, and result in the same spans:
```javascript
router.get('/post/:id', showPost);
router.prefix('/:first');
app.use(router.routes());
```

Let me know if you have any feedback on these span attributes or if this does not match what you were thinking. Also, please let me know if you have any further questions!

Reply:
Great! I was interpreting that as patching the `router.use`, which was incorrect. I should have re-read to double-check.

The name in your screenshot looks like what I'd expect so I expect the rest is legit. Thanks!

I might suggest some tests to represent these sort of cases as example, and to prevent accidentally breaking in the future, but extremely minor suggestion.

Reply:
Looks like I can't mark this resolved, but good on my end.

Reply:
Thanks for the feedback! I appreciate the time you took to review this. I'll add a few tests on these to help in the future!